ouroboros-consensus-test/bench-storage/Bench/Ouroboros/Consensus/Storage/LedgerDB/HD.hs

```diff
@@ -0,0 +1,631 @@
+    })
+
+  pure $ Flush n
+
+-- | Generate a @'Rollback'@ command.
+--
+-- >  data Cmd (i :: Imp) k v =
+-- >    -- ...
+-- >    | Rollback Int (DiffSeq i k v)
+-- >    -- ...
+--
+-- Steps to generate a @'Rollback'@ command:
+-- * Pick how much to rollback, i.e. an integer @n@.
+--   * With high probability, we perform small rollbacks of around @0.05 * k@.
+--   * With low probability, we perform large rollbacks of around @0.95 * k@.
+-- * Return the current diff sequence and @n@.
+-- BOOKKEEPING: Remove the last @n@ diffs from the models' diff sequence.
+--
+-- Note: We assume that pushes do not skip any slots.
+genRollback :: Eq v => GenConfig -> CmdGen v (Cmd 'New Int v)
+genRollback GenConfig{securityParameter = k, rollbackConfig} = do
+    ds <- gets diffs
+    t  <- gets tip
+
+    let
+      RollbackConfig{distribution = dist} =  rollbackConfig
+
+    m <- frequency' [(x, pure $ ofK y) | (x, y) <- dist]
+
+    let
+      n      = min m $ DS.length ds
+      (l, _r) = DS.splitAtFromEnd n ds
+
+    modify (\st -> st {
+        diffs = l
+      , tip = t - n
+      , nrGenerated = nrGenerated st + 1
+      })
+
+    pure $ Rollback n
+  where
+    ofK :: Float -> Int
+    ofK p =
+      let
+        x :: Float
+        x = p * fromIntegral k
+      in
+        round x
+
+-- | Generate a @'Forward'@ command.
+--
+-- >  data Cmd (i :: Imp) k v =
+-- >    -- ...
+-- >    | Forward (Values i k v) (Keys i k v) (DiffSeq i k v)
+--
+-- Steps to generate a @'Forward'@ command:
+-- * Determine which keys to forward.
+-- * Retrieve (a subset of) the backing values from the model, i.e.,
+--   /read/ the keys (obtained in previous step) from the backing values.
+-- * Retrieve the current diff sequence from the model.
+-- * Return the previous three combined.
+-- BOOKKEEPING: None, since forwarding doesn't alter any model state.
+--
+-- Note: The keys to forward should be a superset of the keys of the backing
+-- values that we retrieve from the model.
+genForward :: Eq v => GenConfig -> CmdGen v (Cmd 'New Int v)
+genForward conf = do
+  let
+    ForwardConfig{nrToForward} = forwardConfig conf
+
+  ds  <- gets diffs
+  bvs <- gets backingValues
+
+  let
+    d              = DS.cumulativeDiff ds
+    TT.TableKeys s = TT.diffKeys d
+
+  ksList <- take nrToForward <$> lift (shuffle (toList s))
+
+  let
+    ks = TT.TableKeys . Set.fromList $ ksList
+    vs = TT.restrictValues bvs ks
+
+  modify (\st -> st{
+      nrGenerated = nrGenerated st + 1
+    })
+
+  pure $ Forward vs ks
+
+{-------------------------------------------------------------------------------
+  Sanity checks
+-------------------------------------------------------------------------------}
+
+-- | Sanity checks for interpretation as fold.
+sanityCheck ::
+     ( Eq k, Eq v, Show k, Show v
+     , Isomorphism (DiffSeq i k v) (DS.DiffSeq 'TS.UTxO k v)
+     )
+  => Model k v
+  -> [Cmd i k v]
+  -> DiffSeq i k v
+  -> Property
+sanityCheck model _cmds result = diffs model === to result
```

Comment:
One sanity check that could/should be included is one that checks whether the model and interpretation match for every step of the interpretation function